File: Retaguarda/Torneio.Domain/Entities/ProdutoExtraMembro.cs
namespace Torneio.Domain.Entities;

public class ProdutoExtraMembro
{
    public Guid Id { get; private set; }
    public Guid TorneioId { get; private set; }
    public Guid ProdutoExtraTorneioId { get; private set; }
    public Guid MembroId { get; private set; }
    public decimal Quantidade { get; private set; }
    public decimal ValorCobrado { get; private set; }
    public string? Observacao { get; private set; }
    public bool Ativo { get; private set; }
    public DateTime CriadoEm { get; private set; }

    private ProdutoExtraMembro() { }

    public static ProdutoExtraMembro Criar(
        Guid torneioId,
        Guid produtoExtraTorneioId,
        Guid membroId,
        decimal quantidade,
        decimal valorCobrado,
        string? observacao = null)
    {
        return new ProdutoExtraMembro
        {
            Id = Guid.NewGuid(),
            TorneioId = torneioId,
            ProdutoExtraTorneioId = produtoExtraTorneioId,
            MembroId = membroId,
            Quantidade = quantidade,
            ValorCobrado = valorCobrado,
            Observacao = string.IsNullOrWhiteSpace(observacao) ? null : observacao.Trim(),
            Ativo = true,
            CriadoEm = DateTime.UtcNow
        };
    }

    public void Atualizar(decimal quantidade, decimal valorCobrado, string? observacao)
    {
        Quantidade = quantidade;
        ValorCobrado = valorCobrado;
        Observacao = string.IsNullOrWhiteSpace(observacao) ? null : observacao.Trim();
    }

    public void Reativar(decimal quantidade, decimal valorCobrado, string? observacao)
    {
        Ativo = true;
        Atualizar(quantidade, valorCobrado, observacao);
    }

    public void Desativar() => Ativo = false;
}

File: App/torneio_app/lib/core/models/produto_extra_membro.dart
class ProdutoExtraMembro {
  final String id;
  final String torneioId;
  final String produtoExtraTorneioId;
  final String membroId;
  final String nomeMembro;
  final double quantidade;
  final double valorCobrado;
  final String? observacao;
  final bool ativo;
  final String? parcelaId;
  final bool pago;
  final DateTime? dataPagamento;
  final bool inadimplente;

  const ProdutoExtraMembro({
    required this.id,
    required this.torneioId,
    required this.produtoExtraTorneioId,
    required this.membroId,
    required this.nomeMembro,
    required this.quantidade,
    required this.valorCobrado,
    this.observacao,
    required this.ativo,
    this.parcelaId,
    required this.pago,
    this.dataPagamento,
    required this.inadimplente,
  });

  factory ProdutoExtraMembro.fromJson(Map<String, dynamic> json) => ProdutoExtraMembro(
        id: json['id'] as String,
        torneioId: json['torneioId'] as String,
        produtoExtraTorneioId: json['produtoExtraTorneioId'] as String,
        membroId: json['membroId'] as String,
        nomeMembro: json['nomeMembro'] as String? ?? '',
        quantidade: (json['quantidade'] as num?)?.toDouble() ?? 1,
        valorCobrado: (json['valorCobrado'] as num?)?.toDouble() ?? 0,
        observacao: json['observacao'] as String?,
        ativo: json['ativo'] as bool? ?? true,
        parcelaId: json['parcelaId'] as String?,
        pago: json['pago'] as bool? ?? false,
        dataPagamento: json['dataPagamento'] != null
            ? DateTime.tryParse(json['dataPagamento'] as String)
            : null,
        inadimplente: json['inadimplente'] as bool? ?? false,
      );
}
